// The JENOVA Cognitive Architecture - Protocol Buffer Definitions
// Copyright (c) 2024, orpheus497. All rights reserved.
//
// The JENOVA Cognitive Architecture is licensed under the MIT License.

syntax = "proto3";

package jenova;

// Service definition for distributed JENOVA operations
service JenovaRPC {
  // LLM inference operations
  rpc GenerateText(GenerateRequest) returns (GenerateResponse);
  rpc GenerateTextStream(GenerateRequest) returns (stream GenerateStreamResponse);

  // Embedding operations
  rpc EmbedText(EmbedRequest) returns (EmbedResponse);
  rpc EmbedTextBatch(EmbedBatchRequest) returns (EmbedBatchResponse);

  // Memory operations
  rpc SearchMemory(MemorySearchRequest) returns (MemorySearchResponse);
  rpc SearchMemoryFederated(MemorySearchRequest) returns (MemorySearchResponse);

  // Health and status
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
  rpc GetCapabilities(CapabilitiesRequest) returns (CapabilitiesResponse);
  rpc GetMetrics(MetricsRequest) returns (MetricsResponse);
}

// Message definitions

message GenerateRequest {
  string prompt = 1;
  float temperature = 2;
  float top_p = 3;
  int32 max_tokens = 4;
  map<string, string> metadata = 5;
  string request_id = 6;
}

message GenerateResponse {
  string text = 1;
  int32 tokens_generated = 2;
  double generation_time_seconds = 3;
  string model_name = 4;
  string request_id = 5;
  bool success = 6;
  string error_message = 7;
}

message GenerateStreamResponse {
  string token = 1;
  bool is_final = 2;
  string request_id = 3;
}

message EmbedRequest {
  string text = 1;
  string model_name = 2;
  string request_id = 3;
}

message EmbedResponse {
  repeated float embedding = 1;
  int32 dimension = 2;
  double embedding_time_seconds = 3;
  string request_id = 4;
  bool success = 5;
  string error_message = 6;
}

message EmbedBatchRequest {
  repeated string texts = 1;
  string model_name = 2;
  string request_id = 3;
}

message EmbedBatchResponse {
  repeated EmbeddingResult embeddings = 1;
  double total_time_seconds = 2;
  string request_id = 3;
  bool success = 4;
  string error_message = 5;
}

message EmbeddingResult {
  repeated float embedding = 1;
  int32 index = 2;
}

message MemorySearchRequest {
  string query = 1;
  int32 max_results = 2;
  repeated string memory_types = 3;  // episodic, semantic, procedural
  bool include_content = 4;  // Privacy: false = only embeddings
  string request_id = 5;
}

message MemorySearchResponse {
  repeated MemoryResult results = 1;
  double search_time_seconds = 2;
  string request_id = 3;
  bool success = 4;
  string error_message = 5;
}

message MemoryResult {
  string memory_type = 1;
  string content = 2;  // Empty if include_content=false
  float similarity_score = 3;
  map<string, string> metadata = 4;
  bytes embedding = 5;  // Provided if include_content=false
}

message HealthCheckRequest {
  string requester_id = 1;
}

message HealthCheckResponse {
  string status = 1;  // healthy, degraded, unhealthy
  double cpu_percent = 2;
  double memory_percent = 3;
  double gpu_memory_percent = 4;
  int64 active_requests = 5;
  int64 total_requests_served = 6;
  double uptime_seconds = 7;
  map<string, string> details = 8;
}

message CapabilitiesRequest {
  string requester_id = 1;
}

message CapabilitiesResponse {
  string instance_id = 1;
  string instance_name = 2;

  // Hardware capabilities
  HardwareCapabilities hardware = 3;

  // Model capabilities
  repeated ModelInfo models = 4;

  // Resource sharing settings
  ResourceSharing sharing = 5;

  // Version info
  string jenova_version = 6;
  string protocol_version = 7;
}

message HardwareCapabilities {
  int32 cpu_cores = 1;
  int32 cpu_threads = 2;
  int64 total_ram_mb = 3;
  int64 available_ram_mb = 4;

  repeated GPUInfo gpus = 5;

  string platform = 6;  // linux, darwin, windows
  string architecture = 7;  // x86_64, arm64, etc.
}

message GPUInfo {
  string name = 1;
  string device_type = 2;  // cuda, opencl, vulkan, metal
  int64 total_memory_mb = 3;
  int64 available_memory_mb = 4;
  int32 compute_capability = 5;
}

message ModelInfo {
  string model_name = 1;
  string model_type = 2;  // llm, embedding
  int32 context_size = 3;
  int32 gpu_layers = 4;
  int64 model_size_mb = 5;
  bool available = 6;
}

message ResourceSharing {
  bool share_llm = 1;
  bool share_embeddings = 2;
  bool share_memory = 3;
  int32 max_concurrent_requests = 4;
  int32 current_load_percent = 5;
}

message MetricsRequest {
  string requester_id = 1;
  bool include_history = 2;
}

message MetricsResponse {
  int64 total_requests = 1;
  int64 successful_requests = 2;
  int64 failed_requests = 3;

  double avg_response_time_ms = 4;
  double p95_response_time_ms = 5;
  double p99_response_time_ms = 6;

  int64 total_tokens_generated = 7;
  double avg_tokens_per_second = 8;

  int64 total_embeddings_generated = 9;
  double avg_embedding_time_ms = 10;

  map<string, int64> requests_by_type = 11;

  // Historical data (optional)
  repeated MetricSnapshot history = 12;
}

message MetricSnapshot {
  int64 timestamp = 1;
  double cpu_percent = 2;
  double memory_percent = 3;
  double gpu_memory_percent = 4;
  int64 active_requests = 5;
}
